parsley-debug/shared/src/main/scala/parsley/debugger/ParseAttempt.scala

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright 2020 Parsley Contributors <https://github.com/j-mie6/Parsley/graphs/contributors>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+package parsley.debugger
+
+/** A representation of the attempts a parser has made during parse-time.
+  *
+  * @param rawInput   The input parsed, as raw text.
+  * @param fromOffset This offset is where the parse attempt started in the input.
+  * @param toOffset   This offset is where the parse attempt finished in the input.
+  * @param fromPos    [[fromOffset]] represented as a (line, column) pair.
+  * @param toPos      [[toOffset]] represented as a (line, column pair).
+  * @param success    Was this parse attempt successful?
+  * @param result     If this parse attempt was successful, what did it return?
+  *                   It is guaranteed that `result.isDefined` is true if and only if the attempt
+  *                   is successful.
+  */
+case class ParseAttempt private [parsley] (
```

Comment:
This is public API, is that intentional? If so we need to be really careful with the forwards binary-compatibility. `case class` is _really_ bad for ensuring that an API is forwards evolvable. We can never add a field to this, effectively. Might be worth thinking about the best way to represent this instead.

Reply:
This is exposed through `DebugTree` via `parseResults`. I suppose what could work is to create a trait `ParseAttempt` which exposes the same `val`s / `def`s as this class, and turn this case class into an internal type (possibly `private [parsley]` or `private [debugger]`)?

Reply:
Well, is there really much benefit to having it as a `case class`?

Reply:
Not very much, in [2fd03fe](https://github.com/j-mie6/parsley/pull/219/commits/2fd03fed496771a0a888b6469a94fb3e98a277e0)
 it was changed over to a standard `class`, which can have fields added without breaking the public API, as the minimal information now is exposed via a separate (`sealed`) trait.